Add Terraform migration guide for Unified Phone Experience (#1281)
auth0/docs-v2: main/docs/customize/phone-messages/unified-phone/migrate-to-unified-phone-experience-with-terraform.mdx

description: Learn how to migrate from the legacy Guardian phone configuration to the Unified Phone Experience using the Auth0 Terraform Provider.
---

Migrate your tenant from the legacy [Auth0 Guardian](/docs/secure/multi-factor-authentication/auth0-guardian) phone configuration to the [Unified Phone Experience (UPE)](/docs/customize/phone-messages/unified-phone/configure-unified-phone) using the Auth0 [Terraform Provider](https://registry.terraform.io/providers/auth0/auth0/latest/docs).

After migration, all [multi-factor authentication (MFA) phone notifications](/docs/customize/phone-messages/unified-phone/use-auth0s-unified-phone-experience-for-multi-factor-authentication) route through a single tenant-level phone provider.

## Prerequisites

Before you begin migration, you need:

- [Auth0 Terraform Provider v1.49.0](https://registry.terraform.io/providers/auth0/auth0/latest) or later.
- If you want to configure a custom phone provider, create and deploy a [`custom-phone-provider`](/docs/customize/phone-messages/configure-phone-messaging-providers/configure-a-custom-phone-provider) Auth0 Action to handle message delivery.

> ## Enable the auth0_phone_provider resource and remove legacy configuration

Reviewer: ```suggestion
## Enable the phone provider resource and remove legacy configuration
```

To enable the Terraform `auth0_phone_provider` resource and clean up legacy configuration:

* [Set the phone_consolidated_experience flag to true](#set-the-phone-consolidated-experience-flag-to-true)
* [Add the tenant-level phone provider](#add-the-tenant-level-phone-provider)
* [Add the auth0_branding_phone_notification_template resource (optional)](#add-the-auth0-branding-phone-notification-template-resource-optional)
* [Clean up the auth0_guardian resource phone block](#clean-up-the-auth0-guardian-resource-phone-block)
* [Plan and apply](#plan-and-apply)

> ### Set the phone_consolidated_experience flag to true

Reviewer: ```suggestion
### Set the consolidated experience flag
```

<Tabs>
  <Tab title="Twilio">

```hcl wrap lines
resource "auth0_phone_provider" "default" {
    name = "twilio"

    configuration {
        delivery_methods = ["text", "voice"]
        default_from     = "YOUR_DEFAULT_PHONE_NUMBER"
        sid              = var.twilio_sid
        mssid            = var.twilio_messaging_service_sid # Optional
    }

    credentials {
        auth_token = var.twilio_auth_token
    }
}
```

To send SMS only, set `delivery_methods` to `["text"]`.

  </Tab>
  <Tab title="Custom (Actions)">

```hcl wrap lines
resource "auth0_phone_provider" "default" {
    name = "custom"

    configuration {
        delivery_methods = ["text"]
    }
}
```

For a [custom phone provider](https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/phone_provider), Auth0 expects the [`custom-phone-provider`](/docs/customize/phone-messages/configure-phone-messaging-providers/configure-a-custom-phone-provider) Action to be deployed before you run `terraform apply`. The Action handles delivery.

  </Tab>
</Tabs>

> ### Add the auth0_branding_phone_notification_template resource (optional)

Reviewer: ```suggestion
### Customize OTP message text (optional)
```

If you use `enrollment_message` or `verification_message` on the legacy Guardian phone block to customize one-time passcode (OTP) message text, you can replicate that behavior with the [`auth0_branding_phone_notification_template`](https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/branding_phone_notification_template) resource.

```hcl wrap lines
resource "auth0_branding_phone_notification_template" "default" {
    # See the Auth0 Terraform Provider documentation for available template variables.
}
```

> ### Clean up the auth0_guardian resource phone block

Reviewer: ```suggestion
### Clean up the Auth0 Guardian resource phone block
```

Remove all legacy provider-specific attributes from the [`auth0_guardian`](https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/guardian) phone block. 

> After cleanup, the block should contain only the `enabled` and `message_types` attributes plus a `depends_on` block that references the `auth0_tenant.main` and `auth0_phone_provider.default` resources. This ensures UPE and phone-provider configuration are applied before the legacy Guardian phone block.

Reviewer: ```suggestion
After cleanup, the block should contain only the `enabled` and `message_types` attributes plus a `depends_on` block that references the `auth0_tenant.main` and `auth0_phone_provider.default` resources. This applies UPE and phone-provider configuration before the legacy Guardian phone block.
```

---
description: Learn how to migrate from the legacy Guardian phone configuration to the Unified Phone Experience using the Auth0 Terraform Provider.
title: Migrate to Unified Phone Experience with Terraform
---

Migrate your tenant from the legacy [Auth0 Guardian](/docs/secure/multi-factor-authentication/auth0-guardian) phone configuration to the [Unified Phone Experience](/docs/customize/phone-messages/unified-phone/configure-unified-phone) using the Auth0 [Terraform Provider](https://registry.terraform.io/providers/auth0/auth0/latest/docs). 

After migration, all [multi-factor authentication (MFA) phone notifications](/docs/customize/phone-messages/unified-phone/use-auth0s-unified-phone-experience-for-multi-factor-authentication) route through a single tenant-level phone provider.

<Warning>

This migration requires two separate [`terraform apply`](https://developer.hashicorp.com/terraform/cli/commands/apply) operations. 

If you attempt to complete the migration with a single `terraform apply` command, Auth0 APIs return 403 errors because the legacy phone configuration and the Unified Phone Experience flag cannot be active at the same time.

</Warning>

> Follow these steps to Migrate your legacy Guardian phone configuration to the Unified Phone Experince using Auth0 Terraform provider: 

Reviewer: ```suggestion

```

Reviewer: Experience is misspelled. Also, it's a migration guide so you don't need this sentence.

## Prerequisites

> 

Reviewer: ```suggestion
Before you begin migration, you need:
```

- [Auth0 Terraform Provider v1.41.0](https://registry.terraform.io/providers/auth0/auth0/latest) or later.

> 

Reviewer: ```suggestion
- If you want to use a custom phone provider, create and deploy an Auth0 Actions [`custom-phone-provider`](/docs/customize/phone-messages/configure-phone-messaging-providers/configure-a-custom-phone-provider) trigger to handle message delivery.
```

> ## 1. Enable the Terraform `auth0_phone_provider` resource and remove legacy configuration

Reviewer: ```suggestion
## Enable the Terraform `auth0_phone_provider` resource and remove legacy configuration
```

Reviewer: Usually, Quickstarts have numbered headings but we don't need this here with H2 and H3 levels.

> To enable the Terraform `auth0_phone_provider` resource and clean up legacy configuration, you need to:

Reviewer: ```suggestion
To enable the Terraform `auth0_phone_provider` resource and clean up legacy configuration:
```

> * Set the `phone_consolidated_experience` flag to false

Reviewer: ```suggestion
* [Set the `phone_consolidated_experience` flag to false](#set-the-phone_consolidated_experience-flag-to-false)
```

> * Add the tenant-level phone provider

Reviewer: ```suggestion
* [Add the tenant-level phone provider](#add-the-tenant-level-phone-provider)
```

> * (Optional) Add the `auth0_branding_phone_notification_template` resource

Reviewer: ```suggestion
* (Optional) [Add the `auth0_branding_phone_notification_template` resource](#add-the-auth0_branding_phone_notification_template-resource)
```

> * Clean up the [`auth0_guardian`](https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/guardian) phone block

Reviewer: ```suggestion
* [Clean up Auth0 Guardian phone block](#clean-up-auth0-guardian-phone-block)
```

> * Plan and apply

Reviewer: ```suggestion
* [Plan and apply](#plan-and-apply)
```

### Set the `phone_consolidated_experience` flag to false

> To ensure that your tenant does not route MFA phone notifications using the Unified Phone Experience, edit your [`auth0_tenant`](https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/tenant) resource and set the `phone_consolidated_experience` flag to `false`.

Reviewer: I think this is a little confusing. Aren't we supposed to be migrating to this experience?

Author: This step is to ensure that during this process UPE is disabled because Auth0 does not support the 2 phone experiences to be enabled at the same time.

```hcl wrap lines
resource "auth0_tenant" "main" {
    # ... other tenant settings ...

    phone_consolidated_experience = false
}
```

### Add the tenant-level phone provider

Create a tenant-level phone provider using the Terraform [`auth0_phone_provider`](https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/phone_provider) resource. This provider handles delivery for all phone-based flows once the Unified Phone Experience is enabled.

<Tabs>
  <Tab title="Twilio">

```hcl wrap lines
resource "auth0_phone_provider" "default" {
    name = "twilio"

    configuration {
        delivery_methods = ["text", "voice"]
        default_from     = "YOUR_DEFAULT_PHONE_NUMBER"
        sid              = var.twilio_sid
        mssid            = var.twilio_messaging_service_sid
    }

    credentials {
        auth_token = var.twilio_auth_token
    }
}
```

To send SMS only, set `delivery_methods` to `["text"]`.

  </Tab>
  <Tab title="Custom Action-based provider">

```hcl wrap lines
resource "auth0_phone_provider" "default" {
    name = "custom"

    configuration {
        delivery_methods = ["text"]
    }
}
```

> For custom providers, Auth0 expects a pre-existing Action with the [`custom-phone-provider`](/docs/customize/phone-messages/configure-phone-messaging-providers/configure-a-custom-phone-provider) trigger already deployed. The Action handles the actual delivery logic. Terraform does not manage the Action's existence as part of the `auth0_phone_provider` resource lifecycle.

Reviewer: Added this to the prerequ section

  </Tab>
</Tabs>

### (Optional) Add the `auth0_branding_phone_notification_template` resource

If you use `enrollment_message` or `verification_message` on the legacy Guardian phone block to customize OTP message text, you can replicate that behavior with the [`auth0_branding_phone_notification_template`](https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/branding_phone_notification_template) resource.

```hcl wrap lines
resource "auth0_branding_phone_notification_template" "default" {
    # Refer to the Auth0 Terraform Provider docs for available template variables
}
```

> ### Clean up the [`auth0_guardian`](https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/guardian) phone block

Reviewer: ```suggestion
### Clean up Auth0 Guardian phone block
```

> Remove all legacy provider specific attributes from the `phone` block. After cleanup, the block should contain only `enabled` and `message_types`.

Reviewer: ```suggestion
Remove all legacy provider specific attributes from the [`auth0_guardian`](https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/guardian) phone block. After cleanup, the block should contain only `enabled` and `message_types`.
```

**Before:**

```hcl wrap lines
resource "auth0_guardian" "default" {
    phone {
        enabled       = true
        provider      = "twilio"
        message_types = ["sms"]

        options {
            enrollment_message   = "Your enrollment code is {{code}}"
            verification_message = "Your verification code is {{code}}"
        }

        twilio_config {
            sid                   = var.twilio_sid
            auth_token            = var.twilio_auth_token
            from                  = "YOUR_DEFAULT_PHONE_NUMBER"
            messaging_service_sid = var.twilio_messaging_service_sid
        }
    }
}
```

**After:**

```hcl wrap lines
resource "auth0_guardian" "default" {
    phone {
        enabled       = true
        message_types = ["sms"]
    }
}
```

### Plan and apply

Review the plan carefully before applying. You should see:

- `auth0_phone_provider.default` created
- `auth0_guardian.default` updated (attributes removed)
- `auth0_tenant.main` updated (`phone_consolidated_experience = false`)
- `auth0_branding_phone_notification_template.default` created (if added)

```bash wrap lines
terraform plan
terraform apply
```

> ## 2. Enable the Unified Phone Experience

Reviewer: ```suggestion
## Enable the Unified Phone Experience
```

> Once Step 1 applies cleanly, set [`phone_consolidated_experience`](https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/tenant#phone_consolidated_experience-2) to `true`. This activates the Unified Phone Experience and routes all MFA phone notifications through the `auth0_phone_provider` you configured in Step 1.

Reviewer: ```suggestion
Once you've remove the legacy configuration, use the Terraform provider resource to configure the Auth0 Universal Phone Experience. Set [`phone_consolidated_experience`](https://registry.terraform.io/providers/auth0/auth0/latest/docs/resources/tenant#phone_consolidated_experience-2) to `true`. This activates the Unified Phone Experience and routes all MFA phone notifications through the `auth0_phone_provider` you already configured.
```
